tests/func/test_datachain.py

```diff
@@ -25,34 +33,157 @@ def test_catalog_anon(tmp_dir, catalog, anon):
-    assert DataChain.from_storage(path, catalog=catalog, update=True).count() == 2
+    assert DataChain.from_storage(path, session=test_session).count() == 1
+    assert DataChain.from_storage(path, session=test_session, update=True).count() == 2
+
+
+@pytest.mark.parametrize(
+    "cloud_type,version_aware",
+    [("file", False)],
+    indirect=True,
+)
+def test_from_storage_reindex_expired(tmp_dir, test_session):
+    tmp_dir = tmp_dir / "parquets"
+    os.mkdir(tmp_dir)
+    uri = tmp_dir.as_uri()
+
+    client, path = Client.parse_url(
+        uri, test_session.catalog.cache, **test_session.catalog.client_config
+    )
+    lst_ds_name = listing_dataset_name(client.uri, posixpath.join(path, ""))
 
+    pd.DataFrame({"name": ["Alice", "Bob"]}).to_parquet(tmp_dir / "test1.parquet")
+    assert DataChain.from_storage(uri, session=test_session).count() == 1
+    pd.DataFrame({"name": ["Charlie", "David"]}).to_parquet(tmp_dir / "test2.parquet")
+    # mark dataset as expired
+    test_session.catalog.metastore.update_dataset_version(
+        test_session.catalog.get_dataset(lst_ds_name),
+        1,
+        created_at=datetime.now(timezone.utc) - timedelta(seconds=LISTING_TTL + 20),
+    )
 
-@pytest.mark.parametrize("use_cache", [False, True])
-def test_map_file(cloud_test_catalog, use_cache):
+    # listing was updated because listing dataset was expired
+    assert DataChain.from_storage(uri, session=test_session).count() == 2
+
+
+@pytest.mark.parametrize(
+    "cloud_type",
+    ["s3", "azure", "gs"],
+    indirect=True,
+)
+def test_from_storage_partials(cloud_test_catalog, test_session):
+    def _datasets(session):
+        return sorted(
+            [
+                f"{ds.name}@v{ds.version}"
+                for ds in DataChain.datasets(
+                    session=session, include_listing=True
+                ).collect("dataset")
+                if ds.name.startswith(LISTING_PREFIX)
+            ]
+        )
+
+    ctc = cloud_test_catalog
+    src_uri = ctc.src_uri
+
+    DataChain.from_storage(f"{src_uri}/dogs", session=test_session)
+    assert _datasets(test_session) == [
+        f"{listing_dataset_name(src_uri, 'dogs/')}@v1",
+    ]
+
+    DataChain.from_storage(f"{src_uri}/dogs/others", session=test_session)
+    assert _datasets(test_session) == [
+        f"{listing_dataset_name(src_uri, 'dogs/')}@v1",
+    ]
+
+    DataChain.from_storage(f"{src_uri}", session=test_session)
+    assert _datasets(test_session) == sorted(
+        [
+            f"{listing_dataset_name(src_uri, 'dogs/')}@v1",
+            f"{listing_dataset_name(src_uri, '')}@v1",
+        ]
+    )
+
+    DataChain.from_storage(f"{src_uri}/cats", session=test_session)
+    assert _datasets(test_session) == sorted(
+        [
+            f"{listing_dataset_name(src_uri, 'dogs/')}@v1",
+            f"{listing_dataset_name(src_uri, '')}@v1",
+        ]
+    )
+
+    DataChain.from_storage(f"{src_uri}/cats", session=test_session, update=True)
```

Comment:
I'd suggest to extract the updates to a separate test since these logic can change independently

Reply:
Extracted to separated test case